federation-2/router-bridge/js-src/plan.ts

```diff
@@ -21,63 +47,463 @@ export class BridgeQueryPlanner {
+// instead of GraphQL variables, included randomized alias names, etc. In
+// addition, the servers relied on the fact that dropUnusedDefinitions had been
+// called in order (and that the signature could be parsed as GraphQL) to
+// extract the name of the operation for display. This caused confusion, as the
+// query document shown in the Studio UI wasn't the same as the one actually
+// sent. ApolloServerPluginUsageReporting (previously apollo-engine-reporting)
+// uses a reporting API which requires it to explicitly include the operation
+// name with each signature; this means that the server no longer needs to parse
+// the signature or run its own signature algorithm on it, and the details of
+// the signature algorithm are now up to the reporting agent. That said, not all
+// Studio features will work properly if your signature function changes the
+// signature in unexpected ways.
+//
+// This function used to live in the `apollo-graphql` package in the
+// `apollo-tooling` repository.
+//
+// Note that this is not exactly the same algorithm as the operation registry
+// signature function, which continues to live in `apollo-graphql`.
+import {
+  DirectiveNode,
+  FieldNode,
+  FloatValueNode,
+  FragmentDefinitionNode,
+  FragmentSpreadNode,
+  InlineFragmentNode,
+  IntValueNode,
+  ListValueNode,
+  ObjectValueNode,
+  OperationDefinitionNode,
+  print,
+  SelectionSetNode,
+  separateOperations,
+  StringValueNode,
+  visit,
+} from "graphql";
+import sortBy from "lodash.sortby";
+import { Buffer } from "buffer";
+
+export function defaultUsageReportingSignature(
+  ast: DocumentNode,
+  operationName: string
+): string {
+  return printWithReducedWhitespace(
+    sortAST(
+      removeAliases(hideLiterals(dropUnusedDefinitions(ast, operationName)))
+    )
+  );
+}
+
+// Like the graphql-js print function, but deleting whitespace wherever
+// feasible. Specifically, all whitespace (outside of string literals) is
+// reduced to at most one space, and even that space is removed anywhere except
+// for between two alphanumerics.
+//
+// Note that recent versions of graphql-js contain a stripIgnoredCharacters
+// function; it would be better to use that instead, though whenever we change
+// the signature algorithm it does make every operation appear to change in
+// Studio.
+//
+// In a GraphQL AST (which notably does not contain comments), the only place
+// where meaningful whitespace (or double quotes) can exist is in StringNodes.
+// So to print with reduced whitespace, we:
+// - temporarily sanitize strings by replacing their contents with hex
+// - use the default GraphQL printer
+// - minimize the whitespace with a simple regexp replacement
+// - convert strings back to their actual value We normalize all strings to
+//   non-block strings for simplicity.
+// (Unlike stripIgnoredCharacters, this does not remove commas.)
+function printWithReducedWhitespace(ast: DocumentNode): string {
```

Comment:
this will be fine for now, but maybe let's put a reminder to find a cleaner way sometime :smile:

Reply:
Yes, I ve documented where the code comes from so we know where to refactor it. this code is used in the gateway and in the bridge so we can be sure the usageReportingSignature stays the same when users migrate.